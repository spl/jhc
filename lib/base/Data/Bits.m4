m4_divert(-1)

m4_define(BITSINST,{{

instance Bits $1 where
    x .&. y = and$1 x y
    x .|. y = or$1 x y
    x `xor` y = xor$1 x y
    complement x = complement$1 x
    shiftL x i = shiftL$1 x i
    shiftR x i = shiftR$1 x i

foreign import primitive "And" and$1 :: $1 -> $1 -> $1
foreign import primitive "Or" or$1 :: $1 -> $1 -> $1
foreign import primitive "Xor" xor$1 :: $1 -> $1 -> $1
foreign import primitive "Shr$2" shiftR$1 :: $1 -> Int -> $1
foreign import primitive "Shl" shiftL$1 :: $1 -> Int -> $1
foreign import primitive "Com" complement$1 :: $1 -> $1

}})

m4_divert
