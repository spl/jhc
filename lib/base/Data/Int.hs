{-# OPTIONS_JHC -N #-}
module Data.Int(Int,Int8,Int16,Int32,Int64,IntMax,IntPtr) where


data Int
data Int8
data Int16
data Int32
data Int64
data IntMax
data IntPtr
