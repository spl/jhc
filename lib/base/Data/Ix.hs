module Data.Ix ( Ix(range, index, inRange, rangeSize) ) where


class  Ord a => Ix a  where
    range     :: (a,a) -> [a]
    index     :: (a,a) -> a -> Int
    inRange   :: (a,a) -> a -> Bool
    rangeSize :: (a,a) -> Int

    rangeSize b@(l,h) | null (range b) = 0
                      | otherwise      = index b h + 1
	-- NB: replacing "null (range b)" by  "not (l <= h)"
	-- fails if the bounds are tuples.  For example,
	-- 	(1,2) <= (2,1)
	-- but the range is nevertheless empty
	--	range ((1,2),(2,1)) = []

instance  Ix Char  where
    range (m,n)		= [m..n]
    index b@(c,c') ci
        | inRange b ci  =  fromEnum ci - fromEnum c
        | otherwise     =  error "Ix.index: Index out of range."
    inRange (c,c') i    =  c <= i && i <= c'

instance  Ix Int  where
    range (m,n)		= [m..n]
    index b@(m,n) i
        | inRange b i   =  i - m
        | otherwise     =  error "Ix.index: Index out of range."
    inRange (m,n) i     =  m <= i && i <= n

instance  Ix Integer  where
    range (m,n)		= [m..n]
    index b@(m,n) i
        | inRange b i   =  fromInteger (i - m)
        | otherwise     =  error "Ix.index: Index out of range."
    inRange (m,n) i     =  m <= i && i <= n

-- instance (Ix a,Ix b) => Ix (a, b) -- as derived, for all tuples
-- instance Ix Bool                  -- as derived
-- instance Ix Ordering              -- as derived
-- instance Ix ()                    -- as derived
