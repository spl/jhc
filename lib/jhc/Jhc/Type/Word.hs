module Jhc.Type.Word(module Jhc.Type.Word, module Jhc.Prim.Bits) where

import Jhc.Prim.Bits

-- define the lifted form of the basic
-- numeric types.

data {-# CTYPE "unsigned"  #-} Word = Word Bits32_
data {-# CTYPE "uint8_t"   #-} Word8 = Word8 Bits8_
data {-# CTYPE "uint16_t"  #-} Word16 = Word16 Bits16_
data {-# CTYPE "uint32_t"  #-} Word32 = Word32 Bits32_
data {-# CTYPE "uint64_t"  #-} Word64 = Word64 Bits64_
data {-# CTYPE "uint128_t" #-} Word128 = Word128 Bits128_
data {-# CTYPE "uintptr_t" #-} WordPtr = WordPtr BitsPtr_
data {-# CTYPE "uintmax_t" #-} WordMax = WordMax BitsMax_

data {-# CTYPE "int"      #-} Int = Int Bits32_
data {-# CTYPE "int8_t"   #-} Int8 = Int8 Bits8_
data {-# CTYPE "int16_t"  #-} Int16 = Int16 Bits16_
data {-# CTYPE "int32_t"  #-} Int32 = Int32 Bits32_
data {-# CTYPE "int64_t"  #-} Int64 = Int64 Bits64_
data {-# CTYPE "int128_t" #-} Int128 = Int128 Bits128_
data {-# CTYPE "intptr_t" #-} IntPtr = IntPtr BitsPtr_
data {-# CTYPE "intmax_t" #-} IntMax = IntMax BitsMax_
