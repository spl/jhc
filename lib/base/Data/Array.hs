{-# OPTIONS_JHC -funboxed-tuples #-}
module Data.Array (
    module Data.Ix,  -- export all of Ix
    Array(),
    array,
    listArray,
    accumArray,
    (!),
    bounds,
    indices,
    elems,
    assocs,
    (//),
    accum,
    ixmap
    ) where

import Data.Ix
import Jhc.Int
import Jhc.Prim.Array
import Jhc.Prim.IO

infixl 9  !, //

data Array a b = MkArray !a !a (Array_ b)

array       :: (Ix a) => (a,a) -> [(a,b)] -> Array a b
array b@(s,e) ivs = case newArray (error "array: missing element") (rangeSize b) [(index b x,y) | (x,y) <- ivs] of
        arr -> MkArray s e arr

listArray             :: (Ix a) => (a,a) -> [b] -> Array a b
listArray b vs        =  array b (zipWith (\ a b -> (a,b)) (range b) vs)

(!)                   :: (Ix a) => Array a b -> a -> b
(!) (MkArray s e arr) i =  case unboxInt (index (s,e) i) of i' -> case indexArray__ arr i' of (# r #) -> r

bounds                :: (Ix a) => Array a b -> (a,a)
bounds (MkArray s e _)  =  (s,e)

indices               :: (Ix a) => Array a b -> [a]
indices               =  range . bounds

elems                 :: (Ix a) => Array a b -> [b]
elems a               =  [a!i | i <- indices a]

assocs                :: (Ix a) => Array a b -> [(a,b)]
assocs a              =  [(i, a!i) | i <- indices a]

(//)                  :: (Ix a) => Array a b -> [(a,b)] -> Array a b
a // []               = a
a // new_ivs          = array (bounds a) (old_ivs ++ new_ivs)
                      where
                  	old_ivs = [(i,a!i) | i <- indices a,
                                             i `notElem` new_is]
                  	new_is  = [i | (i,_) <- new_ivs]

accum                 :: (Ix a) => (b -> c -> b) -> Array a b -> [(a,c)] -> Array a b
accum f               =  foldl (\a (i,v) -> a // [(i,f (a!i) v)])

accumArray            :: (Ix a) => (b -> c -> b) -> b -> (a,a) -> [(a,c)] -> Array a b
accumArray f z b      =  accum f (array b [(i,z) | i <- range b])

ixmap                 :: (Ix a, Ix b) => (a,a) -> (a -> b) -> Array b c -> Array a c
ixmap b f a           = array b [(i, a ! f i) | i <- range b]

instance  (Ix a)          => Functor (Array a) where
    fmap fn a = array (bounds a) [ (a,fn b) | (a,b) <- assocs a ]

instance  (Ix a, Eq b)  => Eq (Array a b)  where
    a == a' =  assocs a == assocs a'

instance  (Ix a, Ord b) => Ord (Array a b)  where
    a <= a' =  assocs a <= assocs a'

instance  (Ix a, Show a, Show b) => Show (Array a b)  where
    showsPrec p a = showParen (p > arrPrec) (
                    showString "array " .
                    showsPrec (arrPrec+1) (bounds a) . showChar ' ' .
                    showsPrec (arrPrec+1) (assocs a)                  )
instance  (Ix a, Read a, Read b) => Read (Array a b)  where
    readsPrec p = readParen (p > arrPrec)
           (\r -> [ (array b as, u)
                  | ("array",s) <- lex r,
                    (b,t)       <- readsPrec (arrPrec+1) s,
                    (as,u)      <- readsPrec (arrPrec+1) t ])

-- Precedence of the 'array' function is that of application itself
arrPrec :: Int
arrPrec = 10

foreign import primitive newWorld__ :: a -> World__

newArray :: a -> Int -> [(Int,a)] -> Array_ a
newArray init n xs = case unboxInt n of
    n' -> case newWorld__ (init,n,xs) of
     w -> case newArray__ n' init w of
      (# w, arr #) -> let
        f :: MutArray_ a -> World__ -> [(Int,a)] -> World__
        f arr w [] = w
        f arr w ((i,v):xs) = case unboxInt i of i' -> case writeArray__ arr i' v w of w -> f arr w xs
            in case f arr w xs of w -> Array_ arr `worldDep_` w

foreign import primitive "dependingOn" worldDep_ :: Array_ b -> World__ -> Array_ b
