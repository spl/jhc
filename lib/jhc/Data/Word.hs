{-# LANGUAGE NoImplicitPrelude #-}
module Data.Word(Word,Word8,Word16,Word32,Word64,WordMax,WordPtr) where

import Jhc.Type.Word
