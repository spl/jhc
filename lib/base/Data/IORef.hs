{-# OPTIONS_JHC -funboxed-values -fno-prelude -funboxed-tuples -fffi #-}
module Data.IORef(
    IORef(),	      -- abstract, instance of: Eq
    newIORef,	      -- :: a -> IO (IORef a)
    readIORef,	      -- :: IORef a -> IO a
    writeIORef,	      -- :: IORef a -> a -> IO ()
    modifyIORef,      -- :: IORef a -> (a -> a) -> IO ()
    atomicModifyIORef,-- :: IORef a -> (a -> (a,b)) -> IO b
    ) where

import Jhc.Prim.Array
import Jhc.Basics
import Jhc.IO

data IORef a = IORef (MutArray_ a)

newRef__ a = newArray__ 1# a
writeRef__ m v = writeArray__ m 0# v
readRef__ m = readArray__ m 0#

-- {-# NOINLINE newIORef #-}
newIORef :: a -> IO (IORef a)
newIORef v = fromUIO $ \w -> case newRef__ v w of (# w', r #) -> (# w', IORef r #)

-- {-# NOINLINE readIORef #-}
readIORef :: IORef a -> IO a
readIORef (IORef r) = fromUIO $ \w -> readRef__ r w

-- {-# NOINLINE writeIORef #-}
writeIORef :: IORef a -> a -> IO ()
writeIORef (IORef r) v = fromUIO $ \w -> case writeRef__ r v w of w' -> (# w', () #)

--foreign import primitive eqRef__ :: Ref__ a -> Ref__ a -> Bool

--instance Eq (IORef a) where
--    (IORef x) == (IORef y) = eqRef__ x y

--{-# NOINLINE modifyIORef #-}
modifyIORef :: IORef a -> (a -> a) -> IO ()
modifyIORef (IORef ref) f = fromUIO $ \w -> case readRef__ ref w of
    (# w', a #) -> case writeRef__ ref (f a) w' of
        w'' -> (# w'', () #)

--{-# NOINLINE atomicModifyIORef #-}
atomicModifyIORef :: IORef a -> (a -> (a,b)) -> IO b
atomicModifyIORef (IORef r) f = fromUIO $ \w -> case readRef__ r w of
    (# w', a #) -> case f a of
        (a',b) -> case writeRef__ r a' w' of
            w'' -> (# w'', b #)
