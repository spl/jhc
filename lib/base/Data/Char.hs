module Data.Char (
    isAscii, isLatin1, isControl, isPrint, isSpace, isUpper, isLower,
    isAlpha, isDigit, isOctDigit, isHexDigit, isAlphaNum,
    digitToInt, intToDigit,
    toUpper, toLower,
    ord, chr,
    readLitChar, showLitChar, lexLitChar,

        -- ...and what the Prelude exports
    Char, String
    ) where

--import Array         -- Used for character name table.
import Numeric (readDec, readOct, lexDigits, readHex)
import Prelude.Text
import Jhc.Basics
import Data.Word(Word())

-- Character-testing operations
isAscii, isLatin1, isControl, isPrint, isSpace, isUpper, isLower,
 isAlpha, isDigit, isOctDigit, isHexDigit, isAlphaNum :: Char -> Bool

isAscii c                =  c < '\x80'

isLatin1 c               =  c <= '\xff'

isControl c              =  c < ' ' || c >= '\DEL' && c <= '\x9f'

isPrint c               =  isLatin1 c && not (isControl c)

isSpace c                =  c `elem` " \t\n\r\f\v\xA0"

isUpper c                =  c >= 'A' && c <= 'Z'

isLower c                =  c >= 'a' && c <= 'z'

isAlpha c                =  isUpper c || isLower c

isDigit c                =  c >= '0' && c <= '9'

isOctDigit c             =  c >= '0' && c <= '7'

isHexDigit c             =  isDigit c || c >= 'A' && c <= 'F' ||
                                         c >= 'a' && c <= 'f'

isAlphaNum c             =  isAlpha c || isDigit c


-- Digit conversion operations
digitToInt :: Char -> Int
digitToInt c
  | isDigit c            =  ord c - ord '0'
  | c >= 'a' && c <= 'f' =  ord c - (ord 'a' + 10)
  | c >= 'A' && c <= 'F' =  ord c - (ord 'A' + 10)
  | otherwise            =  error "Char.digitToInt: not a digit"

intToDigit :: Int -> Char
intToDigit i = f (fromIntegral i :: Word) where
    f w | w < 10 = chr (ord '0' + i)
        | w < 16 = chr ((ord 'a' - 10) + i)
        | otherwise = error "Char.intToDigit: not a digit"


-- Case-changing operations
toUpper :: Char -> Char
toUpper c | isLower c = chr $ ord c - 32
          | otherwise = c

toLower :: Char -> Char
toLower c | isUpper c = chr $ ord c + 32
          | otherwise = c


-- Text functions
readLitChar          :: ReadS Char
readLitChar ('\\':s) =  readEsc s
readLitChar (c:s)    =  [(c,s)]

readEsc          :: ReadS Char
readEsc ('a':s)  = [('\a',s)]
readEsc ('b':s)  = [('\b',s)]
readEsc ('f':s)  = [('\f',s)]
readEsc ('n':s)  = [('\n',s)]
readEsc ('r':s)  = [('\r',s)]
readEsc ('t':s)  = [('\t',s)]
readEsc ('v':s)  = [('\v',s)]
readEsc ('\\':s) = [('\\',s)]
readEsc ('"':s)  = [('"',s)]
readEsc ('\'':s) = [('\'',s)]
readEsc ('^':(c:s)) | c >= '@' && c <= '_'
                 = [(chr (ord c - ord '@'), s)]
readEsc s@(d:_) | isDigit d
                 = [(chr n, t) | (n,t) <- readDec s]
readEsc ('o':s)  = [(chr n, t) | (n,t) <- readOct s]
readEsc ('x':s)  = [(chr n, t) | (n,t) <- readHex s]
readEsc s@(c:_) | isUpper c
                 = let table = ('\DEL', "DEL") : zip ['\NUL' .. ] asciiTab
                   in case [(c,s') | (c, mne) <- table,
                                     ([],s') <- [match mne s]]
                      of (pr:_) -> [pr]
                         []     -> []
readEsc _        = []

match                         :: (Eq a) => [a] -> [a] -> ([a],[a])
match (x:xs) (y:ys) | x == y  =  match xs ys
match xs     ys               =  (xs,ys)

showLitChar               :: Char -> ShowS
showLitChar c | c > '\DEL' =  showChar '\\' .
                              protectEsc isDigit (shows (ord c))
showLitChar '\DEL'         =  showString "\\DEL"
showLitChar '\\'           =  showString "\\\\"
showLitChar c | c >= ' '   =  showChar c
showLitChar '\a'           =  showString "\\a"
showLitChar '\b'           =  showString "\\b"
showLitChar '\f'           =  showString "\\f"
showLitChar '\n'           =  showString "\\n"
showLitChar '\r'           =  showString "\\r"
showLitChar '\t'           =  showString "\\t"
showLitChar '\v'           =  showString "\\v"
showLitChar '\SO'          =  protectEsc (== 'H') (showString "\\SO")
showLitChar c              =  showString ('\\' : (asciiTab!!ord c))

protectEsc p f             = f . cont
                             where cont s@(c:_) | p c = "\\&" ++ s
                                   cont s             = s

asciiTab :: [String]
asciiTab = --listArray ('\NUL', ' ')
           ["NUL", "SOH", "STX", "ETX", "EOT", "ENQ", "ACK", "BEL",
            "BS",  "HT",  "LF",  "VT",  "FF",  "CR",  "SO",  "SI",
            "DLE", "DC1", "DC2", "DC3", "DC4", "NAK", "SYN", "ETB",
            "CAN", "EM",  "SUB", "ESC", "FS",  "GS",  "RS",  "US",
            "SP"]

lexLitChar          :: ReadS String
lexLitChar ('\\':s) =  map (prefix '\\') (lexEsc s)
        where
          lexEsc :: String -> [(String,String)]
          lexEsc (c:s)     | c `elem` "abfnrtv\\\"'"  = [([c],s)]
          lexEsc ('^':(c:s)) | (c >= '@') && (c <= '_') = [(['^',c],s)]

          -- Numeric escapes
          lexEsc ('o':s)               = [prefix 'o' (span isOctDigit s)]
          lexEsc ('x':s)               = [prefix 'x' (span isHexDigit s)]
          lexEsc s@(d:_)   | isDigit d = [span isDigit s]

          -- Very crude approximation to \XYZ.
          lexEsc s@(c:_)   | isUpper c = [span isCharName s]
          lexEsc _                     = []

          isCharName c   = isUpper c || isDigit c
          prefix c (t,s) = (c:t, s)

lexLitChar (c:s)    =  [([c],s)]
lexLitChar ""       =  []

