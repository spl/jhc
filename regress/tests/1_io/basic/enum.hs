
import Data.Word

data Day = Sunday | Monday | Tuesday | Wednesday | Thursday | Friday | Saturday
    deriving(Eq,Ord,Enum,Show,Bounded)

main = do
    print (False,True)
    print (fromEnum False, fromEnum True)
    print (toEnum 0 :: Bool, toEnum 1 :: Bool)
    print $ (toEnum 3 :: Day)
    print [10 :: Int, 9 .. -12 ]
    print [1, 5 :: Word8 .. 16 ]
    print [100, 93 :: Word8 .. 43  ]
    print (minBound :: Day,maxBound :: Day)
    print [Friday, Thursday  .. ]
    print [Sunday .. ]
    print [Wednesday .. ]
    print [Wednesday .. Saturday]
    print [Friday, Thursday  .. ]
    print [Monday, Tuesday  .. ]
    print [Monday, Wednesday .. ]


