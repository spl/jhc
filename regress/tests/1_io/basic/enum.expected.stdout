(False,True)
(0,1)
(False,True)
Wednesday
[10,9,8,7,6,5,4,3,2,1,0,-1,-2,-3,-4,-5,-6,-7,-8,-9,-10,-11,-12]
[1,5,9,13]
[100,93,86,79,72,65,58,51,44]
(Sunday,Saturday)
[Friday,Thursday,Wednesday,Tuesday,Monday,Sunday]
[Sunday,Monday,Tuesday,Wednesday,Thursday,Friday,Saturday]
[Wednesday,Thursday,Friday,Saturday]
[Wednesday,Thursday,Friday,Saturday]
[Friday,Thursday,Wednesday,Tuesday,Monday,Sunday]
[Monday,Tuesday,Wednesday,Thursday,Friday,Saturday]
[Monday,Wednesday,Friday]
